src/test/fuzz/versionbits.cpp

```diff
@@ -0,0 +1,345 @@
+#include <primitives/block.h>
+#include <versionbits.h>
+
+#include <test/fuzz/FuzzedDataProvider.h>
+#include <test/fuzz/fuzz.h>
+#include <test/fuzz/util.h>
+
+#include <cstdint>
+#include <limits>
+#include <memory>
+#include <vector>
+
+namespace {
+class TestConditionChecker : public AbstractThresholdConditionChecker
+{
+private:
+    mutable ThresholdConditionCache m_cache;
+    const Consensus::Params dummy_params{};
+
+public:
+    const int64_t m_begin = 0;
+    const int64_t m_end = 0;
+    const int m_period = 0;
+    const int m_threshold = 0;
+    const int m_bit = 0;
+
+    TestConditionChecker(int64_t begin, int64_t end, int period, int threshold, int bit)
+        : m_begin{begin}, m_end{end}, m_period{period}, m_threshold{threshold}, m_bit{bit}
+    {
+        assert(m_period > 0);
+        assert(0 <= m_threshold && m_threshold <= m_period);
+        assert(0 <= m_bit && m_bit <= 32 && m_bit < VERSIONBITS_NUM_BITS);
+    }
+
+    bool Condition(const CBlockIndex* pindex, const Consensus::Params& params) const override { return Condition(pindex->nVersion); }
+    int64_t BeginTime(const Consensus::Params& params) const override { return m_begin; }
+    int64_t EndTime(const Consensus::Params& params) const override { return m_end; }
+    int Period(const Consensus::Params& params) const override { return m_period; }
+    int Threshold(const Consensus::Params& params) const override { return m_threshold; }
+
+    ThresholdState GetStateFor(const CBlockIndex* pindexPrev) const { return AbstractThresholdConditionChecker::GetStateFor(pindexPrev, dummy_params, m_cache); }
+    int GetStateSinceHeightFor(const CBlockIndex* pindexPrev) const { return AbstractThresholdConditionChecker::GetStateSinceHeightFor(pindexPrev, dummy_params, m_cache); }
+    BIP9Stats GetStateStatisticsFor(const CBlockIndex* pindexPrev) const { return AbstractThresholdConditionChecker::GetStateStatisticsFor(pindexPrev, dummy_params); }
+
+    bool Condition(int64_t version) const
+    {
+        return ((version >> m_bit) & 1) != 0 && (version & VERSIONBITS_TOP_MASK) == VERSIONBITS_TOP_BITS;
+    }
+
+    bool Condition(const CBlockIndex* pindex) const { return Condition(pindex->nVersion); }
+};
+
+/** Track blocks mined for test */
+class Blocks
+{
+private:
+    std::vector<std::unique_ptr<CBlockIndex>> m_blocks;
+    const uint32_t m_start_time;
+    const uint32_t m_interval;
+    const int32_t m_signal;
+    const int32_t m_no_signal;
+
+public:
+    Blocks(uint32_t start_time, uint32_t interval, int32_t signal, int32_t no_signal)
+        : m_start_time{start_time}, m_interval{interval}, m_signal{signal}, m_no_signal{no_signal} {}
+
+    size_t size() const { return m_blocks.size(); }
+
+    CBlockIndex* tip() const
+    {
+        return m_blocks.empty() ? nullptr : m_blocks.back().get();
+    }
+
+    CBlockIndex* mine_block(bool signal)
+    {
+        CBlockHeader header;
+        header.nVersion = signal ? m_signal : m_no_signal;
```

Comment:
Any reason to only allow two different nVersion per test input? On the real network there is more noise. The following diff compiled for me, but feel free to ignore:

```diff
diff --git a/src/test/fuzz/versionbits.cpp b/src/test/fuzz/versionbits.cpp
index a898e2782d..1724bde3dc 100644
--- a/src/test/fuzz/versionbits.cpp
+++ b/src/test/fuzz/versionbits.cpp
@@ -64,11 +64,11 @@ private:
     std::vector<std::unique_ptr<CBlockIndex>> m_blocks;
     const uint32_t m_start_time;
     const uint32_t m_interval;
-    const int32_t m_signal;
-    const int32_t m_no_signal;
+    const std::vector<int32_t> m_signal;
+    const std::vector<int32_t> m_no_signal;
 
 public:
-    Blocks(uint32_t start_time, uint32_t interval, int32_t signal, int32_t no_signal)
+    Blocks(uint32_t start_time, uint32_t interval, std::vector<int32_t> signal, std::vector<int32_t> no_signal)
         : m_start_time{start_time}, m_interval{interval}, m_signal{signal}, m_no_signal{no_signal} {}
 
     size_t size() const { return m_blocks.size(); }
@@ -81,7 +81,7 @@ public:
     CBlockIndex* mine_block(bool signal)
     {
         CBlockHeader header;
-        header.nVersion = signal ? m_signal : m_no_signal;
+        header.nVersion = signal ? m_signal.at(m_blocks.size() % m_signal.size()) : m_no_signal.at(m_blocks.size() % m_no_signal.size());
         header.nTime = m_start_time + m_blocks.size() * m_interval;
         header.nBits = 0x1d00ffff;
 
@@ -126,10 +126,6 @@ FUZZ_TARGET_INIT(versionbits, initialize)
 
     const int64_t block_start_time = fuzzed_data_provider.ConsumeIntegralInRange<uint32_t>(params.GenesisBlock().nTime, MAX_TIME);
 
-    // what values for version will we use to signal / not signal?
-    const int32_t ver_signal = fuzzed_data_provider.ConsumeIntegral<int32_t>();
-    const int32_t ver_nosignal = fuzzed_data_provider.ConsumeIntegral<int32_t>();
-
     // select deployment parameters: bit, start time, timeout
     const int bit = fuzzed_data_provider.ConsumeIntegralInRange<int>(0, VERSIONBITS_NUM_BITS - 1);
 
@@ -168,16 +164,25 @@ FUZZ_TARGET_INIT(versionbits, initialize)
 
     TestConditionChecker checker(start_time, timeout, period, threshold, bit);
 
-    // Early exit if the versions don't signal sensibly for the deployment
-    if (!checker.Condition(ver_signal)) return;
-    if (checker.Condition(ver_nosignal)) return;
-    if (ver_nosignal < 0) return;
-
-    // TOP_BITS should ensure version will be positive
-    assert(ver_signal > 0);
+    // what values for version will we use to signal / not signal?
+    std::vector<int32_t> vver_signal;
+    std::vector<int32_t> vver_nosignal;
+    while (vver_signal.size() < 5) {
+        const auto ver_signal = fuzzed_data_provider.ConsumeIntegral<int32_t>();
+        if (!checker.Condition(ver_signal)) return;
+        // TOP_BITS should ensure version will be positive
+        assert(ver_signal > 0);
+        vver_signal.push_back(ver_signal);
+    }
+    while (vver_nosignal.size() < 5) {
+        const auto ver_nosignal = fuzzed_data_provider.ConsumeIntegral<int32_t>();
+        if (ver_nosignal < 0) return;
+        if (checker.Condition(ver_nosignal)) return;
+        vver_nosignal.push_back(ver_nosignal);
+    }
 
     // Now that we have chosen time and versions, setup to mine blocks
-    Blocks blocks(block_start_time, interval, ver_signal, ver_nosignal);
+    Blocks blocks(block_start_time, interval, vver_signal, vver_nosignal);
 
     /* Strategy:
      *  * we will mine a final period worth of blocks, with

Reply:
Oh, I like that!

I don't think it's much benefit for now though -- without some refactoring, I think it's only checking how the `Condition()` function defined in the fuzzer behaves, not the one that's actually used for consensus.

Reply:
Yeah, good point. The first thing I did when reviewing was to change the Condition function to something else. And the test still passed, turns out this is expected. Can be revisited later.